# Load

```{r}
EcophyCofog::Library(c("EcophyCofog","readr","dplyr","magrittr","ggplot2","reshape2","patchwork","ggpubr","tidyr","forcats","readxl","microeco","file2meco","RColorBrewer","vegan","phyloseq","ClustOfVar"))
```

```{r}
load("../../output/microeco_obj.Rdata")
rm(meco_18s,meco_23s)
site_data_raw <- read.csv("../../ressource/Site_edaphic_data/MappPtsSampled.csv")
site_data_raw %<>%
    mutate(Sample=ifelse(nchar(Sample)<3,paste0("0",Sample),Sample))%>%
    mutate(Sample=ifelse(nchar(Sample)<3,paste0("0",Sample),Sample))%>%
    mutate(Sample=ifelse(grepl("^8",Sample),paste0("0",Sample),Sample))%>% # fix 87a and 87b sample ids
    arrange(Sample)# fix sample ID to match the format of other MAPP datasets
```

# process 16s

## Select sites

```{r}
setdiff(gsub("[A-Z]|-|(?=_).*","",meco_16s$sample_table$sample_id,perl = T),site_data_raw$Sample) # pb samples "042047" "047042" "119"    "166"    "116" 
# what about my duplicates 87/256
meco_16s$sample_table %<>% filter(!Sample%in%c("116","119","166","042047","047042")) 
meco_16s$tidy_dataset()
```



## rm singletons

```{r}
nrow(meco_16s$otu_table) #273894 OTUs
meco_16s$otu_table %<>% filter(rowSums(meco_16s$otu_table)>1)
meco_16s$tidy_dataset()
nrow(meco_16s$otu_table) # 23010
meco_16s$sample_table %<>% mutate(nb_reads=colSums(meco_16s$otu_table),
                                  nb_motus=colSums(meco_16s$otu_table!=0))
```

check OTU reads frequency
```{r}
meco_16s$otu_table%>%
    reframe(counts=rowSums(.))%>%
    data.frame()%>%
    ggplot(aes(x=log(counts)))+
    geom_histogram()+
    theme_classic2()
```


## test filtering ASVs freq
```{r}
prevalence <- c(0,1,2,3,4,5,10,15,20) # set prevalence threshold 0/5/10/15 and 20% of our 207 samples 
n_sites <- 207 # 207 samples
prevalence_nsites <- floor(n_sites*0.01*prevalence) # convert to number of sites
relab_thrshld <- c(0,1e-3,1e-5,1e-7)

com_matrices_list <- NULL
for(i in prevalence_nsites){
    matrix_tmp <- t(meco_16s$otu_table)   
    
    matrix_tmp <- matrix_tmp[,which(colSums(matrix_tmp!=0)>i)]
    for(j in relab_thrshld){
        matrix_tmp <- matrix_tmp[,which(colSums(matrix_tmp)/sum(matrix_tmp)>j)]
        matrix_tmp <- matrix_tmp[rowSums(matrix_tmp)!=0,]
        rownames(matrix_tmp) <- gsub("[A-Z]|-|(?=_).*","",rownames(matrix_tmp),perl = T)
        com_matrices_list[[paste0("com_prev_",i,"_relab_",j)]] <- matrix_tmp
    }
}


com_matrices_list <- com_matrices_list[grepl("relab_0$",names(com_matrices_list))]

nmds_test <- NULL
centro_coord <- data.frame(nmds1=character(0),nmds2=character(0),filt=character(0))
for(i in 1:length(com_matrices_list)){
    to_fortify <- metaMDS(com_matrices_list[[i]],distance = "horn")
    nmds_test[[names(com_matrices_list[i])]] <- to_fortify
    centro_coord_tmp <- ggvegan:::fortify.metaMDS(to_fortify)%>%
        filter(score=="sites")%>%
        reframe(nmds1=sum(NMDS1)/202,nmds2=sum(NMDS2)/202)%>%
        mutate(filt=names(com_matrices_list[i]))
    centro_coord <- rbind(centro_coord,centro_coord_tmp) 
}

plot_list <- NULL
for(i in 1:length(nmds_test)){
    data_tmp <- filter(fortify(nmds_test[[i]]),score=="sites")
    data_tmp %<>% mutate(sample=gsub("[A-Z]|-|(?=_).*","",label,perl = T))
    data_tmp$country <- meco_16s$sample_table$Country
    plot_list[[names(nmds_test)[i]]]<- data_tmp %>%
        ggplot(aes(x=NMDS1,y=NMDS2,color=country))+
        geom_point()+
        theme_classic2()+
        ggtitle(paste0(names(nmds_test)[i],": ",nrow(fortify(nmds_test[[i]]))-202," ASVs"))
}
yikes<- (plot_list$com_prev_0_relab_0+plot_list$com_prev_2_relab_0)/(plot_list$com_prev_4_relab_0+plot_list$com_prev_6_relab_0)/(plot_list$com_prev_8_relab_0+plot_list$com_prev_10_relab_0)/(plot_list$com_prev_20_relab_0+plot_list$com_prev_31_relab_0)/(plot_list$com_prev_41_relab_0+plot_spacer())+plot_layout(guides="collect")

procrustes_res <- NULL
for(i in 2:length(nmds_test)){
    procrustes_res[[i-1]] <- procrustes(nmds_test[[1]],nmds_test[[i]])
}


pro_plot <-  NULL
for(i in 1:length(procrustes_res)){
    pro.test <- procrustes_res[[i]]
    ctest <- data.frame(nmds1=pro.test$Yrot[,1],
                        nmds2=pro.test$Yrot[,2],
                        xnmds1=pro.test$X[,1],
                        xnmds2=pro.test$X[,2])
    
    pro.error <- data.frame(err=residuals(procrustes_res[[i]]),sam_id=1:202)
    quantiles <- quantile(pro.error$err,probs = c(0.25,0.50,0.75))
    pro_plot[[i]] <- ggplot(ctest) +
        geom_point(aes(x=nmds1, y=nmds2)) +
        geom_point(aes(x=xnmds1, y=xnmds2)) +
        geom_segment(aes(x=nmds1,y=nmds2,xend=xnmds1,yend=xnmds2),arrow=arrow(length=unit(0.2,"cm")))+
        coord_fixed()+
        theme_classic2()+
        ggtitle(paste0(names(nmds_test)[i+1],": ",nrow(fortify(nmds_test[[i+1]]))-202," ASVs")) +
        ggplot(pro.error,aes(x = sam_id, y=err))+
        geom_bar(stat='identity',width = 1)+
        geom_hline(yintercept = quantiles[1],linetype="dashed")+
        geom_hline(yintercept = quantiles[2])+
        geom_hline(yintercept = quantiles[3],linetype="dashed")+
        theme_classic2()+
        xlab("site row")+
        ylab("Residual fit")+ 
        plot_layout(widths = c(1.4,1.6))
}

ggsave(plot=pro_plot[[1]]/pro_plot[[2]]/pro_plot[[3]]/pro_plot[[4]]/pro_plot[[5]]/pro_plot[[6]]/pro_plot[[7]]/pro_plot[[8]],
       filename = "../../Figures/procrustes.png",width = 10,height = 18)
```


get site coord in our 5 nmds
```{r}
site_coord_nmds <- data.frame(sample=rep(data_tmp$sample,9),
                              data_set=rep(names(nmds_test),each=202),
                              nmds1=c(fortify(nmds_test$com_prev_0_relab_0)$NMDS1[1:202],
                                      fortify(nmds_test$com_prev_2_relab_0)$NMDS1[1:202],
                                      fortify(nmds_test$com_prev_4_relab_0)$NMDS1[1:202],
                                      fortify(nmds_test$com_prev_6_relab_0)$NMDS1[1:202],
                                      fortify(nmds_test$com_prev_8_relab_0)$NMDS1[1:202],
                                      fortify(nmds_test$com_prev_10_relab_0)$NMDS1[1:202],
                                      fortify(nmds_test$com_prev_20_relab_0)$NMDS1[1:202],
                                      fortify(nmds_test$com_prev_31_relab_0)$NMDS1[1:202],
                                      fortify(nmds_test$com_prev_41_relab_0)$NMDS1[1:202]),
                              nmds2=c(fortify(nmds_test$com_prev_0_relab_0)$NMDS2[1:202],
                                      fortify(nmds_test$com_prev_2_relab_0)$NMDS2[1:202],
                                      fortify(nmds_test$com_prev_4_relab_0)$NMDS2[1:202],
                                      fortify(nmds_test$com_prev_6_relab_0)$NMDS2[1:202],
                                      fortify(nmds_test$com_prev_8_relab_0)$NMDS2[1:202],
                                      fortify(nmds_test$com_prev_10_relab_0)$NMDS2[1:202],
                                      fortify(nmds_test$com_prev_20_relab_0)$NMDS2[1:202],
                                      fortify(nmds_test$com_prev_31_relab_0)$NMDS2[1:202],
                                      fortify(nmds_test$com_prev_41_relab_0)$NMDS2[1:202])
                              )


site_coord_nmds %>%
    group_by(sample,data_set)%>%
    mutate(dis_2_centro=sqrt(nmds1^2+nmds2^2))%>%
    pivot_wider(id_cols = sample,
                names_from = data_set,
                values_from = dis_2_centro)%>%
    mutate(delta_0_2=abs(com_prev_0_relab_0-com_prev_2_relab_0),
           delta_2_4=abs(com_prev_2_relab_0-com_prev_4_relab_0),
           delta_4_6=abs(com_prev_4_relab_0-com_prev_6_relab_0),
           delta_6_8=abs(com_prev_6_relab_0-com_prev_8_relab_0),
           delta_8_10=abs(com_prev_8_relab_0-com_prev_10_relab_0),
           delta_10_20=abs(com_prev_10_relab_0-com_prev_20_relab_0),
           delta_20_31=abs(com_prev_20_relab_0-com_prev_31_relab_0),
           delta_31_41=abs(com_prev_31_relab_0-com_prev_41_relab_0))%>%
    pivot_longer(cols=(which(grepl("delta",colnames(.)))),
                 names_to = "evaluated_diff",
                 values_to = "values")%>%
    mutate(evaluated_diff=fct_relevel(evaluated_diff,"delta_0_2","delta_2_4","delta_4_6","delta_6_8","delta_8_10","delta_10_20","delta_20_31","delta_31_41"))%>%
    ggplot(aes(x=evaluated_diff,
               y=values))+
    geom_boxplot(fill="darkorange")+
    theme_classic2()+
    ggtitle("Effect of filtering on site NMDS distance to origin")+
    ylab("Distance to origin absolute differences")+
site_coord_nmds %>%
    group_by(sample,data_set)%>%
    mutate(dis_2_centro=sqrt(nmds1^2-(sum(nmds1)/202)^2+nmds2^2-(sum(nmds2)/202)^2))%>%
    pivot_wider(id_cols = sample,
                names_from = data_set,
                values_from = dis_2_centro)%>%
    mutate(delta_0_2=abs(com_prev_0_relab_0-com_prev_2_relab_0),
           delta_2_4=abs(com_prev_2_relab_0-com_prev_4_relab_0),
           delta_4_6=abs(com_prev_4_relab_0-com_prev_6_relab_0),
           delta_6_8=abs(com_prev_6_relab_0-com_prev_8_relab_0),
           delta_8_10=abs(com_prev_8_relab_0-com_prev_10_relab_0),
           delta_10_20=abs(com_prev_10_relab_0-com_prev_20_relab_0),
           delta_20_31=abs(com_prev_20_relab_0-com_prev_31_relab_0),
           delta_31_41=abs(com_prev_31_relab_0-com_prev_41_relab_0))%>%
    pivot_longer(cols=(which(grepl("delta",colnames(.)))),
                 names_to = "evaluated_diff",
                 values_to = "values")%>%
    mutate(evaluated_diff=fct_relevel(evaluated_diff,"delta_0_2","delta_2_4","delta_4_6","delta_6_8","delta_8_10","delta_10_20","delta_20_31","delta_31_41"))%>%
    ggplot(aes(x=evaluated_diff,
               y=values))+
    geom_boxplot(fill="cornflowerblue")+
    theme_classic2()+
    ggtitle("Effect of filtering on site NMDS distance to centroid")+
    ylab("Distance to centroid absolute differences")
```


```{r}
site_coord_nmds %>%
    group_by(sample,data_set)%>%
    mutate(dis_2_centro=sqrt(nmds1^2+nmds2^2))%>%
    pivot_wider(id_cols = sample,
                names_from = data_set,
                values_from = dis_2_centro)%>%
    mutate(delta_0_2=abs(com_prev_0_relab_0-com_prev_2_relab_0),
           delta_0_4=abs(com_prev_0_relab_0-com_prev_4_relab_0),
           delta_0_6=abs(com_prev_0_relab_0-com_prev_6_relab_0),
           delta_0_8=abs(com_prev_0_relab_0-com_prev_8_relab_0),
           delta_0_10=abs(com_prev_0_relab_0-com_prev_10_relab_0),
           delta_0_20=abs(com_prev_0_relab_0-com_prev_20_relab_0),
           delta_0_31=abs(com_prev_0_relab_0-com_prev_31_relab_0),
           delta_0_41=abs(com_prev_0_relab_0-com_prev_41_relab_0))%>%
    pivot_longer(cols=(which(grepl("delta",colnames(.)))),
                 names_to = "evaluated_diff",
                 values_to = "values")%>%
    mutate(evaluated_diff=fct_relevel(evaluated_diff,"delta_0_2","delta_0_4","delta_0_6","delta_0_8","delta_0_10","delta_0_20","delta_0_31","delta_0_41"))%>%
    ggplot(aes(x=evaluated_diff,
               y=values))+
    geom_boxplot(fill="darkorange")+
    theme_classic2()+
    ggtitle("Effect of filtering on site NMDS distance to origin")+
    ylab("Distance to origin absolute differences")+
site_coord_nmds %>%
    group_by(sample,data_set)%>%
    mutate(dis_2_centro=sqrt(nmds1^2-(sum(nmds1)/202)^2+nmds2^2-(sum(nmds2)/202)^2))%>%
    pivot_wider(id_cols = sample,
                names_from = data_set,
                values_from = dis_2_centro)%>%
    mutate(delta_0_2=abs(com_prev_0_relab_0-com_prev_2_relab_0),
           delta_0_4=abs(com_prev_0_relab_0-com_prev_4_relab_0),
           delta_0_6=abs(com_prev_0_relab_0-com_prev_6_relab_0),
           delta_0_8=abs(com_prev_0_relab_0-com_prev_8_relab_0),
           delta_0_10=abs(com_prev_0_relab_0-com_prev_10_relab_0),
           delta_0_20=abs(com_prev_0_relab_0-com_prev_20_relab_0),
           delta_0_31=abs(com_prev_0_relab_0-com_prev_31_relab_0),
           delta_0_41=abs(com_prev_0_relab_0-com_prev_41_relab_0))%>%
    pivot_longer(cols=(which(grepl("delta",colnames(.)))),
                 names_to = "evaluated_diff",
                 values_to = "values")%>%
    mutate(evaluated_diff=fct_relevel(evaluated_diff,"delta_0_2","delta_0_4","delta_0_6","delta_0_8","delta_0_10","delta_0_20","delta_0_31","delta_0_41"))%>%
    ggplot(aes(x=evaluated_diff,
               y=values))+
    geom_boxplot(fill="cornflowerblue")+
    theme_classic2()+
    ggtitle("Effect of filtering on site NMDS distance to centroid")+
    ylab("Distance to centroid absolute differences")+
    site_coord_nmds %>%
    group_by(sample,data_set)%>%
    mutate(dis_2_centro=sqrt(nmds1^2-(sum(nmds1)/202)^2+nmds2^2-(sum(nmds2)/202)^2))%>%
    mutate(data_set=as.factor(data_set))%>%
    mutate(data_set=fct_relevel(data_set,"com_prev_0_relab_0","com_prev_2_relab_0","com_prev_4_relab_0","com_prev_6_relab_0","com_prev_8_relab_0","com_prev_10_relab_0","com_prev_20_relab_0","com_prev_31_relab_0","com_prev_41_relab_0"))%>%
    ggplot(aes(x=data_set,
               y=dis_2_centro))+
    geom_boxplot(fill="darkolivegreen2")+
    theme_classic2()+
    ggtitle("Distance to centroid in our datasets")+
    ylab("Distance to centroid")
    
```

```{r}
data.frame(RVs=c(FactoMineR::coeffRV(fortify(nmds_test$com_prev_0_relab_0)[1:202,3:4],fortify(nmds_test$com_prev_2_relab_0)[1:202,3:4])$rv,
                 FactoMineR::coeffRV(fortify(nmds_test$com_prev_0_relab_0)[1:202,3:4],fortify(nmds_test$com_prev_4_relab_0)[1:202,3:4])$rv,
                 FactoMineR::coeffRV(fortify(nmds_test$com_prev_0_relab_0)[1:202,3:4],fortify(nmds_test$com_prev_6_relab_0)[1:202,3:4])$rv,
                 FactoMineR::coeffRV(fortify(nmds_test$com_prev_0_relab_0)[1:202,3:4],fortify(nmds_test$com_prev_8_relab_0)[1:202,3:4])$rv,
                 FactoMineR::coeffRV(fortify(nmds_test$com_prev_0_relab_0)[1:202,3:4],fortify(nmds_test$com_prev_10_relab_0)[1:202,3:4])$rv,
                 FactoMineR::coeffRV(fortify(nmds_test$com_prev_0_relab_0)[1:202,3:4],fortify(nmds_test$com_prev_20_relab_0)[1:202,3:4])$rv,
                 FactoMineR::coeffRV(fortify(nmds_test$com_prev_0_relab_0)[1:202,3:4],fortify(nmds_test$com_prev_31_relab_0)[1:202,3:4])$rv,
                 FactoMineR::coeffRV(fortify(nmds_test$com_prev_0_relab_0)[1:202,3:4],fortify(nmds_test$com_prev_41_relab_0)[1:202,3:4])$rv),
           x=1:8)%>%
    ggplot()+
    geom_segment(aes(x=x,y=0,xend=x,yend=RVs),lineend = "square",linewidth=10,color="coral")+
    theme_classic2()+
    ylab("RVs")+
    xlab("Diff from original dataset")
```



## plot pairwise betadiversity

```{r}
store_dist <- NULL
for(i in 1:length(com_matrices_list)){
    dist_tmp <- hillR::hill_taxa_parti_pairwise(com_matrices_list[[i]])
    dist_tmp$filt <- names(com_matrices_list)[i]
    store_dist <- rbind(store_dist,dist_tmp)
}

store_dist%>%
    mutate(filt=as.factor(filt))%>%
    mutate(filt=fct_relevel(filt,"com_prev_0_relab_0","com_prev_2_relab_0","com_prev_4_relab_0","com_prev_6_relab_0","com_prev_8_relab_0","com_prev_10_relab_0","com_prev_20_relab_0","com_prev_31_relab_0","com_prev_41_relab_0"))%>%
    ggplot(aes(x=filt,y=TD_beta))+
    geom_boxplot()+
    theme_classic2()
```


## Rarefaction Curves

```{r}
plot_list_raref <- NULL
for(i in 2:length(com_matrices_list)){
tab16s <- com_matrices_list[[i]] # get the community
curve16s <- rarecurve(tab16s) # use vegan rarefaction curves
names(curve16s) <- rownames(tab16s) # name curves after sample IDs

# Coerce data into "long" form.
protox <- mapply(FUN = function(x, y) {
    mydf <- as.data.frame(x)
    colnames(mydf) <- "value"
    mydf$samples <- y
    mydf$subsample <- attr(x, "Subsample")
    mydf
}, x = curve16s, y = as.list(names(curve16s)), SIMPLIFY = FALSE)

xy <- do.call(rbind, protox)
rownames(xy) <- NULL  # pretty

# Plot.
plot_list_raref[[names(com_matrices_list)[i]]] <- ggplot(xy, aes(x = subsample, y = value, group = samples )) +
    theme_bw() +
    scale_color_discrete(guide = "none") +  # turn legend on or off
    geom_line(color = "aquamarine4")+
    xlab("nb reads")+
    ylab("OTUs count")+
    ggtitle(paste0(nrow(tab16s)," ASVs"))
}
```



## agreggate @genus level

```{r}
phylos_16s <- meco2phyloseq(meco_16s)
phylo_aggclass_16s <- phyloseq::tax_glom(phylos_16s,taxrank = "Class_conf_rdp")
meco_16s_aggclass <- phyloseq2meco(phylo_aggclass_16s)
```


```{r}
com_matrices_list <- NULL
for(i in prevalence_nsites){
    matrix_tmp <- t(meco_16s_aggclass$otu_table)   
    
    matrix_tmp <- matrix_tmp[,which(colSums(matrix_tmp!=0)>i)]
    for(j in relab_thrshld){
        matrix_tmp <- matrix_tmp[,which(colSums(matrix_tmp)/sum(matrix_tmp)>j)]
        matrix_tmp <- matrix_tmp[rowSums(matrix_tmp)!=0,]
        com_matrices_list[[paste0("com_prev_",i,"_relab_",j)]] <- matrix_tmp
    }
}


com_matrices_list <- com_matrices_list[grepl("relab_0$",names(com_matrices_list))]

nmds_test <- NULL
for(i in 1:length(com_matrices_list)){
    nmds_test[[names(com_matrices_list[i])]] <- fortify(metaMDS(com_matrices_list[[i]],
                        distance = "horn"))
    
}

plot_list <- NULL
for(i in 1:length(nmds_test)){
    data_tmp <- filter(nmds_test[[i]],score=="sites")
    data_tmp %<>% mutate(sample=gsub("[A-Z]|-|(?=_).*","",label,perl = T))
    data_tmp$country <- meco_16s$sample_table$Country
    plot_list[[names(nmds_test)[i]]]<- data_tmp %>%
        ggplot(aes(x=NMDS1,y=NMDS2,color=country))+
        geom_point()+
        theme_classic2()+
        ggtitle(paste0(names(nmds_test)[i],": ",nrow(nmds_test[[i]])-207," ASVs"))
}
yikes2<- (plot_list$com_prev_0_relab_0+plot_list$com_prev_10_relab_0)/(plot_list$com_prev_20_relab_0+plot_list$com_prev_31_relab_0)/(plot_list$com_prev_41_relab_0+plot_spacer())+plot_layout(guides="collect")
```


# Process site variables

## Select sites

```{r}
site_data <- site_data_raw %>% filter(Sample%in%meco_16s$sample_table$Sample)%>%
    mutate(across(.cols= colnames(site_data_raw),.fns = ~replace(., . ==  -9999 , NA)))
```


## Count NAs and Imputations

```{r}
var_lt <- colnames(site_data)[grep('_lt', names(site_data))] # extract long term variables (as they also exist on monthly basis)
months_we_want <- c("03","04","05") # march/april/may  (1 to 4 months before sampling)
var_st <- colnames(site_data)[grep('03|04|05', names(site_data))]

site_data_lt <- site_data%>%
    select(all_of(c("X","Y",var_lt,"bulk","gHM","height","ph","soc","water")))
site_data_st <- site_data%>%
    select(all_of(c("X","Y",var_st,"bulk","gHM","height","ph","soc","water")))

as.data.frame(colSums(is.na(site_data)))%>%
    rename(nb_na=`colSums(is.na(site_data))`)%>%
    mutate(var=rownames(.))%>%
    filter(nb_na!=0)%>%
    ggplot(aes(x=var,y=nb_na))+
    geom_bar(stat='identity',fill='midnightblue')+
    theme_classic2()+
    theme(axis.text.x = element_text(angle=45,hjust=1),
          plot.title = element_text(face="bold"))+
    ggtitle("Number of missing values")+
    as.data.frame(colSums(is.na(site_data)))%>%
    rename(nb_na=`colSums(is.na(site_data))`)%>%
    mutate(var=rownames(.),
           na_pct=(nb_na/202)*100)%>%
    filter(nb_na!=0)%>%
    ggplot(aes(x=var,y=na_pct))+
    geom_bar(stat='identity',fill='mediumvioletred')+
    theme_classic2()+
    theme(axis.text.x = element_text(angle=45,hjust=1),
          plot.title = element_text(face="bold"))+
    ggtitle("Percentage of missing values")
```

```{r}
ncp_data_lt <- site_data %>% 
    select(all_of(c("X","Y",var_lt,"bulk","gHM","height","ph","soc","water"))) %>%
    missMDA::estim_ncpPCA(scale=T)


imputed_data_lt <- site_data %>% 
    select(all_of(c("X","Y",var_lt,"bulk","gHM","height","ph","soc","water")))%>%
    missMDA::imputePCA(ncp=5,scale=T)
imputed_data_lt <- imputed_data_lt$completeObs %>% as.data.frame()

ncp_data_st <- site_data %>% 
    select(all_of(c("X","Y",var_st,var_lt,"bulk","gHM","height","ph","soc","water"))) %>%
    mutate(across(everything(),as.numeric))%>%
    missMDA::estim_ncpPCA(scale=T)


imputed_data_st <- site_data %>% 
    select(all_of(c("X","Y",var_st,var_lt,"bulk","gHM","height","ph","soc","water")))%>%
    missMDA::imputePCA(ncp=5,scale=T)
imputed_data_st <- imputed_data_st$completeObs %>% as.data.frame()
```



## Create Short term 

```{r}
site_data_st_before_imputation <- site_data %>% 
    select(all_of(c("X","Y",var_lt,var_st,"bulk","gHM","height","ph","soc","water")))%>%
    mutate(across(.cols=all_of(var_lt),
                  ~get(paste0(gsub("_lt","_",cur_column()), months_we_want[1]))+
                       get(paste0(gsub("_lt","_",cur_column()), months_we_want[2]))+
                       get(paste0(gsub("_lt","_",cur_column()), months_we_want[3]))/3,
                  .names = "{col}__st"
    ))
var_to_remove <- colnames(site_data_st_before_imputation)[grep('_[0-9]{2}', names(site_data_st_before_imputation))]
site_data_st_before_imputation %<>%
    select(-one_of(var_to_remove))%>% 
    rename_with(~ stringr::str_remove(., "_lt_"), everything())

st <- names(site_data_st_before_imputation)[grep('_st', names(site_data_st_before_imputation))]
ncp_data_before_st <- site_data_st_before_imputation %>% 
    select(all_of(c("X","Y",st,"bulk","gHM","height","ph","soc","water"))) %>%
    mutate(across(everything(),as.numeric))%>%
    missMDA::estim_ncpPCA(scale=T)


site_data_st_before_imputation_imputed <- site_data_st_before_imputation %>% 
    select(all_of(c("X","Y",st,"bulk","gHM","height","ph","soc","water")))%>%
    missMDA::imputePCA(ncp=3,scale=T)
site_data_st_before_imputation_imputed <- site_data_st_before_imputation_imputed$completeObs %>% as.data.frame()




site_data_st_after_imputation <- imputed_data_st %>%
    mutate(across(.cols=all_of(var_lt),
                  ~get(paste0(gsub("_lt","_",cur_column()), months_we_want[1]))+
                       get(paste0(gsub("_lt","_",cur_column()), months_we_want[2]))+
                       get(paste0(gsub("_lt","_",cur_column()), months_we_want[3]))/3,
                  .names = "{col}__st"
    ))

var_to_remove <- colnames(site_data_st_after_imputation)[grep('_[0-9]{2}', names(site_data_st_after_imputation))]
site_data_st_after_imputation %<>%
    select(-one_of(var_to_remove))%>% 
    rename_with(~ stringr::str_remove(., "_lt_"), everything())
```







## PCA sites


```{r}
pca_sites_lt <- imputed_data_lt %>% 
    select(!grep('_st', names(imputed_data_lt)))%>%
    FactoMineR::PCA(scale.unit = T,
                    graph = F)

pca_sites_before_st <- site_data_st_before_imputation_imputed %>% 
    select(grep('_st', names(site_data_st_before_imputation_imputed)))%>%
    FactoMineR::PCA(scale.unit = T,
                    graph = F) # NAs

pca_sites_after_st <- site_data_st_after_imputation %>% 
    select(grep('_st', names(site_data_st_after_imputation)))%>%
    FactoMineR::PCA(scale.unit = T,
                    graph = F)
```

```{r}
arrowMul <- function(arrows, data, at = c(0, 0), fill = 0.75) {
  u <- c(range(data[,1], range(data[,2])))
  u <- u - rep(at, each = 2)
  r <- c(range(arrows[, 1], na.rm = TRUE), range(arrows[, 2], na.rm = TRUE))
  rev <- sign(diff(u))[-2]
  if (rev[1] < 0)
    u[1:2] <- u[2:1]
  if (rev[2] < 0)
    u[3:4] <- u[4:3]
  u <- u/r
  u <- u[is.finite(u) & u > 0]
  fill * min(u)
}
```

```{r}
eig <- pca_sites_lt$eig
coord_pca <- as.data.frame(pca_sites_lt$ind$coord)
coord_pca$site <- site_data$Sample
coord_pca$country <- site_data$Country

mul <- arrowMul(as.data.frame(pca_sites_lt$var$coord),
               pca_sites_lt$ind$coord)

pca_site_biplot <- ggplot()+
    geom_point(coord_pca,mapping=aes(x=Dim.1,y=Dim.2),color="darkorange")+
    theme_classic2()+
    geom_vline(xintercept = 0,lty=2)+
    geom_hline(yintercept = 0,lty=2)+
    # ggrepel::geom_label_repel(data=filter(coord_pca,Dim.2< -3),aes(x=Dim.1,y=Dim.2,label=site))+
    xlab(paste0("PC1 (",round(eig[1,2],2),"%)"))+
    ylab(paste0("PC2 (",round(eig[2,2],2),"%)"))+
    geom_segment(data= as.data.frame(pca_sites_lt$var$coord),aes(x = 0, y = 0, xend=mul*Dim.1, yend=mul*Dim.2),
               lineend = "round", 
               linejoin = "round",
               linewidth = .75, 
               arrow = arrow(length = unit(0.2, "inches")),
               colour = "black" 
  )+
    ggrepel::geom_text_repel(data = data.frame(var=rownames(pca_sites_lt$var$coord),pca_sites_lt$var$coord*mul), # add variable names at the end of arrows
            aes(x = ifelse(Dim.1<0,Dim.1*1.1,Dim.1*1.1), # nudge a bit the coordinates so that they're not on the arrows
                y = ifelse(Dim.2<0,Dim.2*1.1,Dim.2*1.1),
                label = var),
             max.overlaps = getOption("ggrepel.max.overlaps", default = 15))+
    ggtitle("Biplot sites long term")+
    theme(plot.title = element_text(face="bold"))
```

```{r}
eig <- pca_sites_before_st$eig
coord_pca <- as.data.frame(pca_sites_before_st$ind$coord)
coord_pca$site <- site_data$Sample
coord_pca$country <- site_data$Country

mul <- arrowMul(as.data.frame(pca_sites_before_st$var$coord),
               pca_sites_before_st$ind$coord)

pca_site_biplot_st_before <- ggplot()+
    geom_point(coord_pca,mapping=aes(x=Dim.1,y=Dim.2),color="cornflowerblue")+
    theme_classic2()+
    geom_vline(xintercept = 0,lty=2)+
    geom_hline(yintercept = 0,lty=2)+
    # ggrepel::geom_label_repel(data=filter(coord_pca,Dim.2< -3),aes(x=Dim.1,y=Dim.2,label=site))+
    xlab(paste0("PC1 (",round(eig[1,2],2),"%)"))+
    ylab(paste0("PC2 (",round(eig[2,2],2),"%)"))+
    geom_segment(data= as.data.frame(pca_sites_before_st$var$coord),aes(x = 0, y = 0, xend=mul*Dim.1, yend=mul*Dim.2),
               lineend = "round", 
               linejoin = "round",
               linewidth = .75, 
               arrow = arrow(length = unit(0.2, "inches")),
               colour = "black" 
  )+
    ggrepel::geom_text_repel(data = data.frame(var=rownames(pca_sites_before_st$var$coord),pca_sites_before_st$var$coord*mul), # add variable names at the end of arrows
            aes(x = ifelse(Dim.1<0,Dim.1*1.1,Dim.1*1.1), # nudge a bit the coordinates so that they're not on the arrows
                y = ifelse(Dim.2<0,Dim.2*1.1,Dim.2*1.1),
                label = var),
             max.overlaps = getOption("ggrepel.max.overlaps", default = 15))+
    ggtitle("Biplot sites short term (mean before imputation)")+
    theme(plot.title = element_text(face="bold"))
```

```{r}
eig <- pca_sites_after_st$eig
coord_pca <- as.data.frame(pca_sites_after_st$ind$coord)
coord_pca$site <- site_data$Sample
coord_pca$country <- site_data$Country

mul <- arrowMul(as.data.frame(pca_sites_after_st$var$coord),
               pca_sites_after_st$ind$coord)

pca_site_biplot_st_after <- ggplot()+
    geom_point(coord_pca,mapping=aes(x=Dim.1,y=Dim.2),color="darkorchid4")+
    theme_classic2()+
    geom_vline(xintercept = 0,lty=2)+
    geom_hline(yintercept = 0,lty=2)+
    # ggrepel::geom_label_repel(data=filter(coord_pca,Dim.2< -3),aes(x=Dim.1,y=Dim.2,label=site))+
    xlab(paste0("PC1 (",round(eig[1,2],2),"%)"))+
    ylab(paste0("PC2 (",round(eig[2,2],2),"%)"))+
    geom_segment(data= as.data.frame(pca_sites_after_st$var$coord),aes(x = 0, y = 0, xend=mul*Dim.1, yend=mul*Dim.2),
               lineend = "round", 
               linejoin = "round",
               linewidth = .75, 
               arrow = arrow(length = unit(0.2, "inches")),
               colour = "black" 
  )+
    ggrepel::geom_text_repel(data = data.frame(var=rownames(pca_sites_after_st$var$coord),pca_sites_after_st$var$coord*mul), # add variable names at the end of arrows
            aes(x = ifelse(Dim.1<0,Dim.1*1.1,Dim.1*1.1), # nudge a bit the coordinates so that they're not on the arrows
                y = ifelse(Dim.2<0,Dim.2*1.1,Dim.2*1.1),
                label = var),
             max.overlaps = getOption("ggrepel.max.overlaps", default = 15))+
    ggtitle("Biplot sites short term (mean on imputed data)")+
    theme(plot.title = element_text(face="bold"))
```

```{r}
pca_site_biplot+pca_site_biplot_st_before+pca_site_biplot_st_after
```


## Cor plot

```{r}
plot_pairs <- GGally::ggpairs(select(site_data,where(is.numeric)),progress = F)
```


Dont like clust of var. Not based on biological rational
```{r}
tree <- ClustOfVar::hclustvar(select(site_data,where(is.numeric)))
plot(tree)
```

```{r}
test <- ClustOfVar::kmeansvar(X.quanti = select(site_data,where(is.numeric)),init=8)
```



```{r}
ClustOfVar::plot.clustvar(test)
```


# Test dbRDAs

```{r}
dbrda_data16s <- site_data%>% 
    select(!grep('_st', names(site_data)))%>%
    select(-c(1:4,17,38))%>%
    mutate(across(.cols=everything() ,.fns = ~replace(., . ==  -9999 , NA)))
dbrda_data16s$nb_reads <- meco_16s$sample_table$nb_reads

com_16s <- com_matrices_list$com_prev_0_relab_0
rownames(com_16s) <- gsub("[A-Z]|-|(?=_).*","",rownames(com_16s),perl = T)
com_16s_hell <- labdsv::hellinger(com_16s)

dbrda16s <- dbrda(com_16s_hell~Condition(nb_reads)+aet_lt+bulk+clay+def_lt+evi_lt+fpar_lt+gHM+gpp_lt+height+lai_lt+moist_lt+ndvi_lt+npp_lt+pdsi_lt+pet_lt+ph+pr_lt+smp_lt+soc+srad_lt+ssm_lt+ssma_lt+susm_lt+susma10+susma_lt+swe_lt+tmmn_lt+tmmx_lt+vap_lt+vpd_lt+water,data=dbrda_data16s, distance="horn") # full model with morisita horn distance based on hellinger transformed data

#get eigen values of axes
eig <- c(dbrda16s$CCA$eig)
eig1 <- round((eig[1]/sum(eig))*100,digits = 2) # % of explained inertia of the first axis
eig2 <- round((eig[2]/sum(eig))*100,digits = 2) # % of explained inertia of the second axis

fort_rda16s <- fortify(dbrda16s) # transform dbrda results in a ggplot2 usable format
vars <- c("dbRDA1","dbRDA2") # we will plot in the first two dimensions
want <- fort_rda16s[["score"]] == "biplot" # get variables scores (arrows)
mul <- arrowMul(fort_rda16s[want, vars, drop = FALSE],
                fort_rda16s[!want, vars, drop = FALSE]) # get the scaling factor to scale these scores for plotting as in ggvegan
fort_rda16s[want, vars] <- mul * fort_rda16s[want, vars] # scale the scores 


veg_dbrda16s_full <- ggplot(fort_rda16s, aes(x = dbRDA1, y = dbRDA2)) +
  ggConvexHull::geom_convexhull(data = cbind(subset(fort_rda16s, score == "sites"), # create convex hull around our treatments
                               Trtmt=filtered_16s_hell$sample_table$Trtmt),
                  aes(fill = Trtmt, color = Trtmt),
                  alpha=0.2)+
  scale_fill_manual(values =mo_col)+ # duplicate colours so it matches substrate colors
  geom_point(data = cbind(subset(fort_rda16s, score == "sites"), # add points (sites) corresponding to samples
                          Trtmt=filtered_16s_hell$sample_table$Trtmt, # add trtmt info for plotting
                          substrate=filtered_16s_hell$sample_table$substrate, # add substrate info
                          drought=filtered_16s_hell$sample_table$drought), # add drought
             aes(color = Trtmt, shape = drought), # color by substrate and shape by drought
             size = 3) +
  # geom_text(aes(label=label))+
  scale_shape_manual(values =  c(17,19))+
  geom_segment(data = subset(fort_rda16s, score == "biplot"), # add arrows for variables
               aes(x = 0, y = 0, xend=dbRDA1, yend=dbRDA2),
               lineend = "round", 
               linejoin = "round",
               linewidth = .75, 
               arrow = arrow(length = unit(0.2, "inches")),
               colour = "black" 
  ) +
  coord_fixed()+ # fix proportion between axis units
  geom_text(data = fort_rda16s[want, , drop = FALSE ], # add variable names at the end of arrows
            aes(x = ifelse(dbRDA1<0,dbRDA1*1.2,dbRDA1*1.2), # nudge a bit the coordinates so that they're not on the arrows
                y = ifelse(dbRDA2<0,dbRDA2*1.2,dbRDA2*1.2),
                label = label))+
  scale_colour_manual(values = mo_col)+
  theme_classic()+
  xlab(paste0("dbRDA1 (",eig1,"%)"))+
  ylab(paste0("dbRDA2 (",eig2,"%)"))+
  guides(fill="none")  +
  geom_hline(yintercept=0,linetype = "dashed")+
  geom_vline(xintercept=0,linetype = "dashed")
```

# Enzymes


```{r}
enzymes_data <- read_excel("../../ressource/enzymes/enzymes_mapp.xlsx")
enzymes_data %<>%
    mutate(ID=ifelse(nchar(ID)<3,paste0("0",ID),ID))%>%
    mutate(ID=ifelse(nchar(ID)<3,paste0("0",ID),ID))%>%
    mutate(ID=ifelse(grepl("^8",ID),paste0("0",ID),ID))%>% # fix 87a and 87b sample ids
    filter(!grepl("246",ID))%>%
    rename(Sample=ID)


```


# Cluster data

## set up
```{r}

library(dendextend)

createAngleHJustCols <- function(labeldf) {        
    nn <- length(labeldf$y)
    halfn <- floor(nn/2)
    firsthalf <- rev(90 + seq(0,360, length.out = nn))
    secondhalf <- rev(-90 + seq(0,360, length.out = nn))
    angle <- numeric(nn)
    angle[1:halfn] <- firsthalf[1:halfn]
    angle[(halfn+1):nn] <- secondhalf[(halfn+1):nn]

    hjust <- numeric(nn)
    hjust[1:halfn] <- 0
    hjust[(halfn+1):nn] <- 1

    return(list(angle = angle, hjust = hjust))
}

```

## 16s
```{r}
com <- labdsv::hellinger(t(meco_16s$otu_table))
rownames(com) <- gsub("[A-Z]|-|(?=_).*","",rownames(com),perl = T)
dist_16s <- vegdist(com,method="horn")
clust_16s <- hclust(dist_16s,method="ward.D2")
hcd_16s <- as.dendrogram(clust_16s)
dend_data_16s <- ggdendro::dendro_data(hcd_16s,type='rectangle')

ang_and_just <- createAngleHJustCols(dend_data_16s$labels)

dend_data_16s$labels %<>% rename(Sample=label)
dend_data_16s$labels %<>% left_join(site_data[,2:3])

plot1 <- ggplot(dend_data_16s$segments)+
    geom_segment(aes(x=x,y=-y,xend=xend,yend=-yend))+
    geom_text(data = dend_data_16s$labels,aes(x,y+0.2,label=Sample,colour=Country),size=3,angle=ang_and_just$angle,hjust=ang_and_just$hjust)+
    theme_void()+
    coord_polar(theta = "x")+
    scale_color_manual(values=c("#7f61d3",
                                "#84b83a",
                                "#be4eb2",
                                "#4db955",
                                "#d54272",
                                "#6abf87",
                                "#d14b33",
                                "#3fc1bf",
                                "#d99037",
                                "#616bba",
                                "#b8af3d",
                                "#c387d1",
                                "#5a8233",
                                "#dd81a0",
                                "#37835d",
                                "#9a4769",
                                "#c6a96d",
                                "#619dd7",
                                "#816e2b",
                                "#bf6a4d"))
```

## sites

```{r}
imputed_data_lt %>% 
    select(!grep('_st', names(imputed_data_lt)))

site_data_st_before_imputation_imputed %>% 
    select(grep('_st', names(site_data_st_before_imputation_imputed)))

site_data_st_after_imputation %>% 
    select(grep('_st', names(site_data_st_after_imputation)))



site_lt_dend <- imputed_data_lt %>% 
    select(!grep('_st', names(imputed_data_lt)))%>%
    select(c(X,Y))
rownames(site_lt_dend) <- site_data$Sample


dist_site_lt <- vegdist(site_lt_dend,method="euclidean")
clust_site_lt <- hclust(dist_site_lt,method="ward.D2")
hcd_site_lt <- as.dendrogram(clust_site_lt)
dend_data_site_lt <- ggdendro::dendro_data(hcd_site_lt,type='rectangle')

ang_and_just <- createAngleHJustCols(dend_data_site_lt$labels)

dend_data_site_lt$labels %<>% rename(Sample=label)
dend_data_site_lt$labels %<>% left_join(site_data[,2:3])


plot2 <- ggplot(dend_data_site_lt$segments)+
    geom_segment(aes(x=x,y=-sqrt(y),xend=xend,yend=-sqrt(yend)))+
    geom_text(data = dend_data_site_lt$labels,aes(x,-(y)+2,label=Sample,color=Country),size=3,angle=ang_and_just$angle,hjust=ang_and_just$hjust)+
    theme_void()+
    coord_polar(theta = "x")+
    scale_color_manual(values=c("#7f61d3",
                                "#84b83a",
                                "#be4eb2",
                                "#4db955",
                                "#d54272",
                                "#6abf87",
                                "#d14b33",
                                "#3fc1bf",
                                "#d99037",
                                "#616bba",
                                "#b8af3d",
                                "#c387d1",
                                "#5a8233",
                                "#dd81a0",
                                "#37835d",
                                "#9a4769",
                                "#c6a96d",
                                "#619dd7",
                                "#816e2b",
                                "#bf6a4d"))
```

## Enzymes

```{r}
ez_dend <- enzymes_data[,-1]
rownames(ez_dend) <- enzymes_data$Sample
dist_ez <- vegdist(ez_dend,method="euclidean")

clust_ez <- hclust(dist_ez,method="ward.D2")
hcd_ez <- as.dendrogram(clust_ez)
dend_data_ez <- ggdendro::dendro_data(hcd_ez,type='rectangle')

ang_and_just <- createAngleHJustCols(dend_data_ez$labels)

dend_data_ez$labels %<>% rename(Sample=label)
dend_data_ez$labels %<>% left_join(site_data[,2:3])


plot3 <- ggplot(dend_data_ez$segments)+
    geom_segment(aes(x=x,y=-(y),xend=xend,yend=-(yend)))+
    geom_text(data = dend_data_ez$labels,aes(x,-(y)+2,label=Sample,color=Country),size=3,angle=ang_and_just$angle,hjust=ang_and_just$hjust)+
    theme_void()+
    coord_polar(theta = "x")+
    scale_color_manual(values=c("#7f61d3",
                                "#84b83a",
                                "#be4eb2",
                                "#4db955",
                                "#d54272",
                                "#6abf87",
                                "#d14b33",
                                "#3fc1bf",
                                "#d99037",
                                "#616bba",
                                "#b8af3d",
                                "#c387d1",
                                "#5a8233",
                                "#dd81a0",
                                "#37835d",
                                "#9a4769",
                                "#c6a96d",
                                "#619dd7",
                                "#816e2b",
                                "#bf6a4d"))
```


```{r}
clust <- plot1+ggtitle("Community clustering")+plot2+ggtitle("Site Long Term clustering")+plot3+ggtitle("Enzymes activities clustering")+plot_layout(guides = "collect")
ggsave("../../Figures/clust.png",width =20,height = 13)
```

# Picrust2


## Alignment and stuff

```{r}
pi_1 <-  read_delim("../../ressource/Picrust2/Galaxy93-[FROGSFUNC_1_placeseqs_and_copynumbers__frogsfunc_placeseqs_closests_ref_sequences.txt].tsv", # read the tsv file
                            delim = "\t", escape_double = FALSE, 
                            trim_ws = TRUE)

pi_1 %>%
    ggplot()+
    geom_point(aes(x=NSTI,y=`%cov`,color="darkorange"))+
    geom_point(aes(x=NSTI,y=`%id`,color="darkorchid4"))+
    theme_classic2()+
    ylab("Metric %") +
     scale_color_identity(name = "Blast metrics",
                          breaks = c("darkorange", "darkorchid4"),
                          labels = c("Coverage", "Identity"),guide = "legend")+ 
    scale_x_continuous(breaks = scales::breaks_width(2),expand = c(0, 0))+
    ggplot(pi_1)+
    geom_point(aes(x=NSTI,y=score), color="aquamarine4")+
    theme_classic2()+
    ylab("Blast score")
```

Clusters with NSTI > 2 are considered poorly represented in the Picrust2 reference database.
Additionnaly, a coverage < 50% is pretty low 

```{r}
(pi_1 %>%
    ggplot(aes(x=score,y=`%id`))+
    geom_point()+
    pi_1 %>%
    ggplot(aes(x=`%id`,y=`%cov`))+
    geom_point()+
    pi_1 %>%
    ggplot(aes(x=score,y=`%cov`))+
    geom_point())/
    (pi_1 %>%
         filter(`%cov`>80&`%id`>80)%>%
    ggplot(aes(x=score,y=`%id`))+
    geom_point()+
    pi_1 %>%
         filter(`%cov`>80&`%id`>80)%>%
    ggplot(aes(x=`%id`,y=`%cov`))+
    geom_point()+
    pi_1 %>%
         filter(`%cov`>80&`%id`>80)%>%
    ggplot(aes(x=score,y=`%cov`))+
    geom_point())

```

```{r}
pi_1 %>%
         filter(`%cov`>80&`%id`>80)%>%
    nrow()/nrow(pi_1) # 93% of OTUs

pi_1 %>%
         filter(`%cov`>80&`%id`>80)%>%
    summarise(sum_seq=sum(`Nb sequences`))/sum(pi_1$`Nb sequences`) # Keep 98% of sequences
```

I don't have much confidence in these OTUs associated predictions cuz poorly mapped in Picrust2 reference db so filter them out.

```{r}
asvs_picrust <- pi_1 %>%
         filter(`%cov`>80&`%id`>80)%>%select(`#ASV`)
asvs_picrust <- asvs_picrust$`#ASV`
```


```{r}
meco_16s_picrust2 <- clone(meco_16s)

meco_16s_picrust2$otu_table %<>% filter(rownames(.)%in%asvs_picrust)
meco_16s_picrust2$tidy_dataset()
```

```{r}
ab_trans_meco_16s_picrust2 <- trans_abund$new(meco_16s_picrust2,taxrank = "Order",ntaxa = 8)
order_relab_16s_picrust2 <- ab_trans_meco_16s_picrust2$plot_bar(others_color = "grey70", xtext_keep = FALSE, legend_text_italic = FALSE)+
    geom_col(width = 1)+
    ggtitle("MAPP 16s")+
    theme(plot.title = element_text(face='bold'))
```

It changes a bit my 8 dominant taxa

```{r}
pi_2 <-  read_delim("../../ressource/Picrust2/Galaxy95-[FROGSFUNC_1_placeseqs_and_copynumbers__frogsfunc_marker.tsv].tsv", # read the tsv file
                            delim = "\t", escape_double = FALSE, 
                            trim_ws = TRUE)
```



## Kegg
```{r}
kegg_pi <- read_delim("../../ressource/Picrust2/Galaxy103-[FROGSFUNC_2_functions__KO_copynumbers_predicted.tsv].tsv", # read the tsv file
                            delim = "\t", escape_double = FALSE, 
                            trim_ws = TRUE)

kegg_pi2 <- read_delim("../../ressource/Picrust2/Galaxy109-[FROGSFUNC_2_functions___frogsfunc_functions_unstrat_KO.tsv].tsv", # read the tsv file
                            delim = "\t", escape_double = FALSE, 
                            trim_ws = TRUE)

kegg_pi3 <- read_delim("../../ressource/Picrust2/Galaxy99-[FROGSFUNC_2_functions__frogsfunc_functions_marker_norm.tsv].tsv", # read the tsv file
                            delim = "\t", escape_double = FALSE, 
                            trim_ws = TRUE)
```

Rows of Kegg_pi2 correspond to non null columns of Kegg_pi 


```{r}
names(kegg_pi2)[grepl("L001",names(kegg_pi2))] <- gsub("[A-Z]|-|(?=_).*","",names(kegg_pi2)[grepl("L001",names(kegg_pi2))] ,perl = T)

kegg_pi2%>%
    select(c(1:4,which(names(.)%in%site_data$Sample)))%>%
    filter(grepl("glucosidase",classification))%>%
    mutate(across(5:ncol(.),~(.x/sum(.))*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value,fill=c4))+
    geom_bar(stat = "identity",position = "stack",width=1)+
    scale_fill_manual(values =RColorBrewer::brewer.pal(11,'Set3'))+
    ylab("Pathway %")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))+
    kegg_pi2%>%
    select(c(1:4,which(names(.)%in%site_data$Sample)))%>%
    filter(grepl("aminopep",classification))%>%
    mutate(across(5:ncol(.),~(.x/sum(.))*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value,fill=c4))+
    geom_bar(stat = "identity",position = "stack",width=1)+
    scale_fill_manual(values =c("#4e8937",
                                "#733cc9",
                                "#6dc140",
                                "#cf4cc3",
                                "#63b889",
                                "#d24172",
                                "#4ca3ae",
                                "#d74c33",
                                "#656ad1",
                                "#cf9133",
                                "#6d2f7d",
                                "#a6a34d",
                                "#bb86cb",
                                "#4a5c2f",
                                "#6f97d0",
                                "#814025",
                                "#434a7b",
                                "#cf8c70",
                                "#78334b",
                                "#d0849f"))+
    ylab("Enzymes")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))
```

```{r}
(EC_pi2%>%
    select(c(1:4,which(names(.)%in%enzymes_data$Sample)))%>%
    # mutate(across(5:ncol(.),~(.x/observation_sum)*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    filter(grepl("leucyl aminopep",classification))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value))+
    geom_bar(stat = "identity",width=1,fill="#d74c33")+
    ylab("Picrust2 leucyl aminopep")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))+
    enzymes_data%>%
    ggplot(aes(x=Sample,y=LAP))+
    geom_bar(stat='identity',fill="#d74c33",width=1)+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0)))/
    (EC_pi2%>%
    select(c(1:4,which(names(.)%in%enzymes_data$Sample)))%>%
    # mutate(across(5:ncol(.),~(.x/observation_sum)*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    filter(grepl("beta-glucosidase",classification))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value))+
    geom_bar(stat = "identity",width=1,fill="#656ad1")+
    ylab("Picrust2 beta-glucosidase")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))+
    enzymes_data%>%
    ggplot(aes(x=Sample,y=BG))+
    geom_bar(stat='identity',fill="#656ad1",width=1)+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0)))

```


```{r}
EC_pi2%>%
    select(c(1:4,which(names(.)%in%mutate(meco_16s$sample_table,name=gsub("[A-Z]|-|(?=_).*","",sample_id ,perl = T))$Sample)))%>%
    # mutate(across(5:ncol(.),~(.x/observation_sum)*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    group_by(name)%>%
    summarise(val=sum(value))%>%
    left_join(mutate(meco_16s$sample_table,name=gsub("[A-Z]|-|(?=_).*","",sample_id ,perl = T)))%>%
    ggplot(aes(val,nb_reads))+
    geom_point()+
    xlab("Picrust2 copy estimation")
```


```{r}
path_EC_pi3 <- read_delim("../../ressource/Picrust2/Galaxy115-[FROGSFUNC_3_pathways__frogsfunc_pathways_unstrat.tsv].tsv", # read the tsv file
                            delim = "\t", escape_double = FALSE, 
                            trim_ws = TRUE)
```

```{r}
names(path_EC_pi3)[grepl("L001",names(path_EC_pi3))] <- gsub("[A-Z]|-|(?=_).*","",names(path_EC_pi3)[grepl("L001",names(path_EC_pi3))] ,perl = T)
```


```{r}
path_EC_pi3%>%
    select(c(1:4,which(names(.)%in%site_data$Sample)))%>%
    mutate(across(5:ncol(.),~(.x/sum(.))*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value,fill=c1))+
    geom_bar(stat = "identity",position = "stack",width=1)+
    scale_fill_manual(values =RColorBrewer::brewer.pal(7,'Dark2'))+
    ylab("Pathway %")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))
    
```

```{r}
path_EC_pi3%>%
    select(c(1:4,which(names(.)%in%site_data$Sample)))%>%
    filter(grepl("Degradation;",classification))%>%
    mutate(across(5:ncol(.),~(.x/sum(.))*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value,fill=c2))+
    geom_bar(stat = "identity",position = "stack",width=1)+
    scale_fill_manual(values =RColorBrewer::brewer.pal(11,'Set3'))+
    ylab("Pathway %")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))
```


## Meta cyc

```{r}
EC_pi <- read_delim("../../ressource/Picrust2/Galaxy102-[FROGSFUNC_2_functions__EC_copynumbers_predicted.tsv].tsv", # read the tsv file
                            delim = "\t", escape_double = FALSE, 
                            trim_ws = TRUE)

EC_pi2 <- read_delim("../../ressource/Picrust2/Galaxy108-[FROGSFUNC_2_functions___frogsfunc_functions_unstrat_EC.tsv].tsv", # read the tsv file
                            delim = "\t", escape_double = FALSE, 
                            trim_ws = TRUE)
```

Rows of EC_pi2 correspond to non null columns of EC_pi 


```{r}
names(EC_pi2)[grepl("L001",names(EC_pi2))] <- gsub("[A-Z]|-|(?=_).*","",names(EC_pi2)[grepl("L001",names(EC_pi2))] ,perl = T)

EC_pi2%>%
    select(c(1:4,which(names(.)%in%site_data$Sample)))%>%
    filter(grepl("glucosidase",classification))%>%
    mutate(across(5:ncol(.),~(.x/sum(.))*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value,fill=c4))+
    geom_bar(stat = "identity",position = "stack",width=1)+
    scale_fill_manual(values =RColorBrewer::brewer.pal(11,'Set3'))+
    ylab("Pathway %")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))+
    EC_pi2%>%
    select(c(1:4,which(names(.)%in%site_data$Sample)))%>%
    filter(grepl("aminopep",classification))%>%
    mutate(across(5:ncol(.),~(.x/sum(.))*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value,fill=c4))+
    geom_bar(stat = "identity",position = "stack",width=1)+
    scale_fill_manual(values =c("#4e8937",
                                "#733cc9",
                                "#6dc140",
                                "#cf4cc3",
                                "#63b889",
                                "#d24172",
                                "#4ca3ae",
                                "#d74c33",
                                "#656ad1",
                                "#cf9133",
                                "#6d2f7d",
                                "#a6a34d",
                                "#bb86cb",
                                "#4a5c2f",
                                "#6f97d0",
                                "#814025",
                                "#434a7b",
                                "#cf8c70",
                                "#78334b",
                                "#d0849f"))+
    ylab("Enzymes")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))
```

```{r}
(EC_pi2%>%
    select(c(1:4,which(names(.)%in%enzymes_data$Sample)))%>%
    # mutate(across(5:ncol(.),~(.x/observation_sum)*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    filter(grepl("leucyl aminopep",classification))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value))+
    geom_bar(stat = "identity",width=1,fill="#d74c33")+
    ylab("Picrust2 leucyl aminopep")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))+
    enzymes_data%>%
    ggplot(aes(x=Sample,y=LAP))+
    geom_bar(stat='identity',fill="#d74c33",width=1)+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0)))/
    (EC_pi2%>%
    select(c(1:4,which(names(.)%in%enzymes_data$Sample)))%>%
    # mutate(across(5:ncol(.),~(.x/observation_sum)*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    filter(grepl("beta-glucosidase",classification))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value))+
    geom_bar(stat = "identity",width=1,fill="#656ad1")+
    ylab("Picrust2 beta-glucosidase")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))+
    enzymes_data%>%
    ggplot(aes(x=Sample,y=BG))+
    geom_bar(stat='identity',fill="#656ad1",width=1)+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0)))

```



```{r}
EC_pi2%>%
    select(c(1:4,which(names(.)%in%enzymes_data$Sample)))%>%
    # mutate(across(5:ncol(.),~(.x/observation_sum)*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    filter(grepl("Bacterial leucyl aminopep",classification))%>%
    group_by(name)%>%
    summarise(val=sum(value))%>%
    left_join(mutate(enzymes_data,name=Sample))%>%
    ggplot(aes(val,LAP))+
    geom_point()+
    xlab("Picrust2 copy estimation")+
    EC_pi2%>%
    select(c(1:4,which(names(.)%in%enzymes_data$Sample)))%>%
    # mutate(across(5:ncol(.),~(.x/observation_sum)*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    filter(grepl("beta-glucosidase",classification))%>%
    group_by(name)%>%
    summarise(val=sum(value))%>%
    left_join(mutate(enzymes_data,name=Sample))%>%
    ggplot(aes(val,BG))+
    geom_point()+
    xlab("Picrust2 copy estimation")
```




```{r}
EC_pi2%>%
    select(c(1:4,which(names(.)%in%mutate(meco_16s$sample_table,name=gsub("[A-Z]|-|(?=_).*","",sample_id ,perl = T))$Sample)))%>%
    # mutate(across(5:ncol(.),~(.x/observation_sum)*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    group_by(name)%>%
    summarise(val=sum(value))%>%
    left_join(mutate(meco_16s$sample_table,name=gsub("[A-Z]|-|(?=_).*","",sample_id ,perl = T)))%>%
    ggplot(aes(val,nb_reads))+
    geom_point()+
    xlab("Picrust2 copy estimation")
```


```{r}
path_kegg_pi3 <- read_delim("../../ressource/Picrust2/Galaxy117-[FROGSFUNC_3_pathways__frogsfunc_pathways_unstrat.tsv].tsv", # read the tsv file
                            delim = "\t", escape_double = FALSE, 
                            trim_ws = TRUE)
```

```{r}
names(path_kegg_pi3)[grepl("L001",names(path_kegg_pi3))] <- gsub("[A-Z]|-|(?=_).*","",names(path_kegg_pi3)[grepl("L001",names(path_kegg_pi3))] ,perl = T)
```


```{r}
path_kegg_pi3%>%
    select(c(1:4,which(names(.)%in%site_data$Sample)))%>%
    mutate(across(5:ncol(.),~(.x/sum(.))*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value,fill=c1))+
    geom_bar(stat = "identity",position = "stack",width=1)+
    scale_fill_manual(values =RColorBrewer::brewer.pal(7,'Dark2'))+
    ylab("Pathway %")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))
    
```

```{r}
path_kegg_pi3%>%
    select(c(1:4,which(names(.)%in%site_data$Sample)))%>%
    filter(grepl("Metabolism;",classification))%>%
    mutate(across(5:ncol(.),~(.x/sum(.))*100))%>%
    pivot_longer(cols =c(5:ncol(.)))%>%
    separate(col=classification,into = paste0("c",1:4),sep = ";")%>%
    ggplot(aes(x=name,y=value,fill=c2))+
    geom_bar(stat = "identity",position = "stack",width=1)+
    scale_fill_manual(values =RColorBrewer::brewer.pal(11,'Set3'))+
    ylab("Pathway %")+
    xlab("Samples")+
    theme_classic2()+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank())+
    scale_y_continuous(expand = c(0, 0))
```




```{r}
BiocManager::install("ggtree")
library(ggtree)
tree <- treeio::read.nhx("../../ressource/Picrust2/Galaxy90-[FROGSFUNC_1_placeseqs_and_copynumbers__frogsfunc_placeseqs_tree.nwk].nhx")
ggtree(tree)
```

# Cytometry